packages/dds/shared-object-base/src/sharedObject.ts

```diff
@@ -343,6 +348,12 @@ export abstract class SharedObjectCore<TEvent extends ISharedObjectEvents = ISha
     protected submitLocalMessage(content: any, localOpMetadata: unknown = undefined): void {
         this.verifyNotClosed();
         if (this.isAttached()) {
+            if (this.runtime.idCompressor !== undefined
+                && content.idRange === undefined) {
```

Comment:
While it would certainly be a moment of "aha, no way!", I could see `idRange` someday colliding with an actual user's unrelated property on their data, and that would be bad and sad.

We can't use a symbol, since those don't get json'd, but we have a pattern elsewhere of using underscores (e.g. for serialized handles). You could match that pattern by using a property called "__fluid_idRange__".

Eventually, the content object gets wrapped in an `envelope`, and it'd be nice if we could put the property on there instead, or some other wrapper (and perhaps it would be delivered to us in the `message` object of `process()` rather than `message.content`).

Mixing a property into the object that is provided by the user feels like a bit of a kludge to me, even if it's a property that's hard to collide with.

Reply:
I'm more inclined to the `__fluid_idRange__` approach here. I think doing this at the `SharedObjectCore` level is correct, but I do agree that it feels somewhat kludgy to stick it in the op's contents field. We could do something a little more complicated, which would be to leverage something like `localOpMetadata` and have it "hoisted" out at some point before the op is sent and added as a property at the `IDocumentMessage` level (I haven't actually tried this, so not sure it would work in practice). I'm not sure how much we gain from doing it that way besides making it more difficult for ourselves as that will require a service change to get the `idRange` round-tripped back to us by the service.

Reply:
That's an interesting idea, although that will have the same problem because it might collide with user's data in `localOpMetadata`, won't it?

What is your rationale for why doing this at the `SharedObjectCore` level is correct? The more I think about it, the more I'm convinced that it should be done at a higher level. After all, this is a single id compressor object which is shared by the entire container. So it seems to me that it makes the most sense for the "container" to handle the ranges. There's no reason to scope it down to the level of the DDS, because it doesn't require any context from the DDS to do its job. A proof of this is that if DDS A allocates 100 ids but doesn't send an op, and then DDS B allocates no IDs but does send an op, then DDS B's op will contain the range with all of the ids from DDS A. That's _weird_, imo. There's nothing technically wrong with it - it should work - but it's _conceptually_ weird to put the code here because it seems like the DDS is responsible for the range that it sends (because that code is inside SharedObjectCore), when really it doesn't necessarily have anything to do with it. It also puts additional constraints on `submitLocalMessage`. Suppose one day we wanted to make `submitLocalMessage` async; if we did, then the ranges could start being submitted out of order.

All that said, if putting stuff higher up in the container runtime is complicated and will block progress, this seems like an okay temporary solution. But it feels like a kludge to me; I'm curious to hear what you think.

Reply:
Your point about having other DDSs ranges in the idRange was actually the motivation for moving it down to the `SharedObjectCore` level, but I realize now that I came to the wrong conclusion when I thought it through. My initial implementation had it higher up but had some issues with replaying pending states when the container's connection state changed. Let me try moving things up and see if I can work through the issues I was running into as I think I'm also convinced this should move higher up.

Reply:
Sounds good! As we discussed offline, having it in SharedObjectCore also presents danger during op resubmission, because after SharedObjectCore has tagged the op contents with the ID range, it might be accidentally removed by a DDS which overrides `reSubmitCore` and tweaks the op object, or submits a different op instead. This would cause the idRange to get dropped on the floor and range finalization to fail shortly thereafter.